tests/evm_backends/base_env.py

```diff
@@ -0,0 +1,228 @@
+from eth.precompiles.point_evaluation import kzg_to_versioned_hash
+from eth_account._utils.typed_transactions.base import TRUSTED_SETUP
+from eth_keys.datatypes import PrivateKey
+from eth_utils import to_checksum_address
+
+from tests.evm_backends.abi import abi_decode
+from tests.evm_backends.abi_contract import ABIContract, ABIContractFactory, ABIFunction
+from vyper.ast.grammar import parse_vyper_source
+from vyper.compiler import CompilerData, Settings, compile_code
+from vyper.utils import ERC5202_PREFIX, method_id
+
+
+# a very simple log representation for the raw log entries
+@dataclass
+class LogEntry:
+    address: str
+    topics: list[str]
+    data: tuple[list[bytes], bytes]  # (topic list, non-topic)
+
+
+# object returned by `last_result` property
+@dataclass
+class ExecutionResult:
+    is_success: bool
+    logs: list[LogEntry]
+    gas_refunded: int
+    gas_used: int
+
+
+class EvmError(RuntimeError):
+    """Exception raised when a transaction reverts."""
+
+
+class BaseEnv:
+    """
+    Base class for EVM backends.
+    It provides a common interface for deploying contracts and interacting with them.
+    """
+
+    INVALID_OPCODE_ERROR = "NotImplemented"  # must be implemented by subclasses
+    DEFAULT_CHAIN_ID = 1
+
+    def __init__(self, gas_limit: int, account_keys: list[PrivateKey]) -> None:
+        self.gas_limit = gas_limit
+        self._keys = account_keys
+        self.deployer: str = self._keys[0].public_key.to_checksum_address()
+
+    @contextmanager
+    def sender(self, address: str):
+        original_deployer = self.deployer
+        self.deployer = address
+        try:
+            yield
+        finally:
+            self.deployer = original_deployer
+
+    def deploy(self, abi: list[dict], bytecode: bytes, value=0, *args, **kwargs):
+        """Deploy a contract with the given ABI and bytecode."""
+        factory = ABIContractFactory.from_abi_dict(abi, bytecode=bytecode)
+
+        initcode = bytecode
+        if args or kwargs:
+            ctor_abi = next(i for i in abi if i["type"] == "constructor")
+            ctor = ABIFunction(ctor_abi, contract_name=factory._name)
+            initcode += ctor.prepare_calldata(*args, **kwargs)
+
+        try:
+            deployed_at = self._deploy(initcode, value)
+        except RuntimeError as e:
+            raise EvmError(*e.args) from e
+
+        address = to_checksum_address(deployed_at)
+        return factory.at(self, address)
+
+    def deploy_source(
+        self,
+        source_code: str,
+        output_formats: dict[str, Callable[[CompilerData], str]],
+        compiler_settings: Settings,
+        *args,
+        input_bundle=None,
+        **kwargs,
+    ) -> ABIContract:
+        """Compile and deploy a contract from source code."""
+        abi, bytecode = _compile(source_code, output_formats, compiler_settings, input_bundle)
+        value = (
+            kwargs.pop("value", 0) or kwargs.pop("value_in_eth", 0) * 10**18
```

Comment:
not sure how i feel about `value_in_eth`, i see it is used in one place and there we should probably be using wei anyways.